typespec-extension/src/code-model-builder.ts

```diff
@@ -2228,81 +2037,44 @@ export class CodeModelBuilder {
+        keyType: {
+          kind: "string",
+          encode: "string",
+          nullable: false,
+        },
+        description: type.description,
+        nullableValues: false,
+        nullable: false,
+        valueType: type.additionalProperties,
+      };
+      const parentSchema = this.processDictionarySchemaFromSdkType(sdkDictType, "Record");
       objectSchema.parents = new Relations();
       objectSchema.parents.immediate.push(parentSchema);
       pushDistinct(objectSchema.parents.all, parentSchema);
-    }
-
-    // value of the discriminator property
-    if (objectSchema.parents) {
-      const parentWithDiscriminator = objectSchema.parents.all.find(
-        (it) => it instanceof ObjectSchema && it.discriminator,
-      );
-      if (parentWithDiscriminator) {
-        discriminatorPropertyName = (
-          (parentWithDiscriminator as ObjectSchema).discriminator as discriminatorTypeWithPropertyName
-        ).propertyName;
-
-        const discriminatorProperty = Array.from(type.properties.values()).find(
-          (it) =>
-            it.name === discriminatorPropertyName &&
-            (it.type.kind === "String" || it.type.kind === "EnumMember" || it.type.kind === "UnionVariant"),
-        );
-        if (discriminatorProperty) {
-          if (discriminatorProperty.type.kind === "String") {
-            // value as StringLiteral
-            objectSchema.discriminatorValue = discriminatorProperty.type.value;
-          } else if (discriminatorProperty.type.kind === "EnumMember") {
-            // value as EnumMember
-            // lint requires value be string
-            objectSchema.discriminatorValue =
-              (discriminatorProperty.type.value as string) ?? discriminatorProperty.type.name;
-          } else if (discriminatorProperty.type.kind === "UnionVariant") {
-            // value as UnionVariant
-            objectSchema.discriminatorValue =
-              ((discriminatorProperty.type.type as StringLiteral).value as string) ?? discriminatorProperty.type.name;
-          }
-        } else {
-          // it is possible that the property is Union, e.g. 'kind: "type1" | "type2"'; but such Type appears not to be a concrete model.
-
-          // fallback to name of the Model
-          objectSchema.discriminatorValue = name;
-        }
-      }
+      objectSchema.discriminatorValue = type.discriminatorValue;
     }
 
     // properties
-    for (const prop of type.properties.values()) {
+    for (const prop of type.properties) {
+      // skip discriminator property
       if (
-        prop.name === discriminatorPropertyName || // skip the discriminator property
-        isNeverType(prop.type) || // skip property of type "never"
-        !isPayloadProperty(this.program, prop)
+        prop.name === type.discriminatorProperty?.name ||
+        prop.name === type.baseModel?.discriminatorProperty?.name ||
+        !isPayloadProperty(this.program, prop.__raw) // why have this check?
```

Comment:
answer the question. tsp Type could include `@header` `@query` `@path` property.

Reply:
Would like to check if we can remove this check now.

Reply:
see https://github.com/Azure/autorest.java/pull/2698#discussion_r1582741965

We should use TCGC info, if available.

Reply:
Yes, I understand we should use TCGC info, rather than `__raw`, I modified the helper functions that used from `processSchema` as much as possible. I did not modify this function because it is also used by `getEffectiveSchemaType`(https://github.com/Azure/autorest.java/blob/main/typespec-extension/src/code-model-builder.ts#L2309), and `getEffectiveSchemaType` is used when processing operation, I have not started on the operation part yet. So I keep it here. 
I can also write another `isPayloadPropertyFromSdkType(prop: SdkModelPropertyType)` and call the function here, then switch to use this function on other places when working on operation part code (switch to `sdkPackage`).
Please let me know your preference.

Reply:
I mean here you do not need to call anything. Just check this `SdkBodyModelPropertyType.discriminator`.
https://github.com/Azure/typespec-azure/blob/main/packages/typespec-client-generator-core/src/interfaces.ts#L367

As in the case of `neverType`, TCGC should have filtered those properties that we previously filter by `!isPayloadProperty`

Reply:
updated